confluent-local phase 1: start, produce/consume and stop using kafka-local image

MuweiHe opened the pull request:
Release Notes
-------------

New Features
* Added new Docker-based `confluent local` commands for improved local testing of Confluent Platform

Bug Fixes
* Remove unused `--output` flag in `confluent kafka topic consume`

Checklist
---------
1. [CRUCIAL] Is the change for features that are already live in prod?
   * yes: ok

What
----
<!--
Briefly describe **what** you have changed and **why**.
Optionally include your implementation strategy.
-->
Delete the old confluent local. Add new commands: start, produce, consume and stop, and list/create/describe/update topics. Start will pull the image and create/start a container with broker and krest configured, with the next available ports. Produce will run a ckgo producer to a topic, consume will consume from a specified topic as well. Stop will remove the container and remove the ports configuration from config.json.

References
----------
<!--
Copy and paste links to tickets, related PRs, GitHub issues, etc.
-->
https://docs.google.com/document/d/1zOi3y2vK3itecv364LTYZzR5gn7Adxx8Ya8nqR8Twy0/edit#

Test & Review
-------------
<!--
Has it been tested? How?
Copy and paste any instructions or steps that can save the reviewer time.
-->
Manually tested. unit and int tests on Linux

Open Questions / Follow-ups
---------------------------
<!--
Optional: Anything open to discussion for the reviewer, out of scope of this PR, or follow-ups.
-->

brianstrauch commented on internal/cmd/kafka/command_topic_consume_onprem.go:
```diff
@@ -52,7 +52,6 @@ func (c *authenticatedTopicCommand) newConsumeCommandOnPrem() *cobra.Command {
 	cmd.Flags().StringSlice("config", nil, `A comma-separated list of configuration overrides ("key=value") for the consumer client.`)
 	cmd.Flags().String("config-file", "", "The path to the configuration file (in json or avro format) for the consumer client.")
 	cmd.Flags().String("schema-registry-endpoint", "", "The URL of the Schema Registry cluster.")
-	pcmd.AddOutputFlag(cmd)
```
This looks like a breaking change. Did you mean to delete this?

brianstrauch replied:
(Also, is it even being used?)

MuweiHe replied:
That's why it's being deleted... we never used this flag

brianstrauch replied:
Can you add this to the "Bug Fixes" section of the release notes?

MuweiHe replied:
will do!

brianstrauch commented:
Can you add release notes?

brianstrauch commented on internal/cmd/local/command.go:
```diff
@@ -9,14 +9,17 @@ import (
 
-type Command struct {
+const imageName = "523370736235.dkr.ecr.us-west-2.amazonaws.com/confluentinc/kafka-local:latest"
+const testTopicName = "jsontest"
```
Would it make more sense to put this constant in command_kafka.go?

brianstrauch commented on internal/cmd/local/command.go:
```diff
@@ -9,14 +9,17 @@ import (
+const imageName = "523370736235.dkr.ecr.us-west-2.amazonaws.com/confluentinc/kafka-local:latest"
+const testTopicName = "jsontest"
+
+type localCommand struct {
```
```suggestion
type command struct {
```

brianstrauch replied:
Do we still need this struct? This command isn't authenticated, and I don't think we're using `ch` or `cc` anymore.

MuweiHe replied:
`cannot define new methods on non-local type cmd.CLICommand` so we still need a struct

MuweiHe replied:
i changed it to `command`

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,23 @@
+)
+
+func NewKafkaCommand(prerunner cmd.PreRunner) *cobra.Command {
+	c := NewLocalCommand(
```
Let's make sure we get rid of this style and copy the style of our other commands.

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,23 @@
+		&cobra.Command{
+			Use:   "kafka",
+			Short: "Run Kafka related commands",
+			Long:  `----`,
```
I take it this PR is not ready for review yet?

MuweiHe replied:
no, it's just a progress tracking for the cp team

brianstrauch replied:
Feel free to mark as "Draft" in that case!

brianstrauch commented:
I assume this command will work on Windows, unlike the old `confluent local`? If that's the case, can we simplify test/help_test.go?

brianstrauch commented on internal/cmd/kafka/command_topic_create_onprem.go:
```diff
@@ -63,10 +63,10 @@ func (c *authenticatedTopicCommand) createOnPrem(cmd *cobra.Command, args []stri
 	}
 
-	return CreateTopicWithRestClient(cmd, restClient, restContext, topicName, clusterId)
+	return CreateTopicWithRESTClient(cmd, restClient, restContext, topicName, clusterId)
```
I think we decided to format functions this way- this helps with readability in certain cases: `TestIamRbac` is easier to read than `TestIAMRBAC` for instance.
```suggestion
	return CreateTopicWithRestClient(cmd, restClient, restContext, topicName, clusterId)
```

brianstrauch commented on commands.txt:
```diff
@@ -0,0 +1,7 @@
+available commands:
```
Is this documentation for the user? Did you intend to upload it or should it belong in a README.md?

MuweiHe replied:
no this was just a note. had deleted it!

brianstrauch commented on commands.txt:
```diff
@@ -0,0 +1,7 @@
+available commands:
+1. start the service with `confluent local kafka start`
```
Any future plans to start all of the services with `confluent local start`? (once we have more than one)

brianstrauch commented on internal/cmd/kafka/kafka_rest_onprem.go:
```diff
@@ -25,14 +26,15 @@ func initKafkaRest(c *pcmd.AuthenticatedCLICommand, cmd *cobra.Command) (*kafkar
 	kafkaRestClient := kafkaREST.Client
-	setServerURL(cmd, kafkaRestClient, url)
+	SetServerURL(cmd, kafkaRestClient, url)
+	fmt.Println("url", url)
```
```suggestion
```

brianstrauch commented on internal/cmd/local/command.go:
```diff
@@ -6,40 +6,21 @@ import (
+	}
 
 	if runtime.GOOS == "windows" {
-		c.Hidden = true
+		cmd.Hidden = true
 	}
```
This works on Windows, right?

MuweiHe replied:
needs testing. TODO @sgagniere

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,30 @@
+	pcmd "github.com/confluentinc/cli/internal/pkg/cmd"
+)
+
+const imageName = "523370736235.dkr.ecr.us-west-2.amazonaws.com/confluentinc/kafka-local:latest"
```
Will there be a public image available soon?

MuweiHe replied:
expected end of q1

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+const imageName = "523370736235.dkr.ecr.us-west-2.amazonaws.com/confluentinc/kafka-local:latest"
+const urlPrefix = "http://0.0.0.0:%s"
```
What is this the URL for? Make this variable name more descriptive if it's global.

MuweiHe replied:
it's for local communication of krest, but might be used later by sr too. `localhostPrefix`?

brianstrauch replied:
Ah, cool. Can you change it to "http://localhost:%d" maybe? (or better yet, use the `url` package!)

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,30 @@
+const imageName = "523370736235.dkr.ecr.us-west-2.amazonaws.com/confluentinc/kafka-local:latest"
+const urlPrefix = "http://0.0.0.0:%s"
+
+type kafkaCommand struct {
+	*pcmd.CLICommand
+}
```
None of these commands should require authentication, so we only need one command `struct` that can be defined in internal/cmd/local/command.go

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,30 @@
+func newKafkaCommand(prerunner pcmd.PreRunner) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "kafka",
+		Short: "Run Kafka related commands",
```
```suggestion
		Short: "Manage a single-node instance of Apache Kafka.",
```

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,30 @@
+	cmd := &cobra.Command{
+		Use:   "kafka",
+		Short: "Run Kafka related commands",
+		Long:  `Run Kafka commands including starting kafka service, produce/consume and stopping the service`,
```
Don't think this is needed since the list of subcommands will be visible here.
```suggestion
```

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,30 @@
+		Use:   "kafka",
+		Short: "Run Kafka related commands",
+		Long:  `Run Kafka commands including starting kafka service, produce/consume and stopping the service`,
+		Args:  cobra.NoArgs,
```
Not needed since this is an internal command.
```suggestion
```

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,30 @@
+
+	cmd.AddCommand(c.newStartCommand())
+	cmd.AddCommand(c.newStopCommand())
+	cmd.AddCommand(c.newTopicCommand())
```
```suggestion
	cmd.AddCommand(c.newTopicCommand())
	
```

MuweiHe replied:
what is the suggested change here? ❓

brianstrauch replied:
Add a newline to separate the subcommands from the return statement!

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,161 @@
+	}
+	defer dockerClient.Close()
+
+	_, err = dockerClient.Info(context.Background())
+	if err != nil {
+		return err
+	}
```
Is this to check if Docker is running? I assume we need this for every command, so why not put this in a `checkIsDockerRunning()` function?

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,161 @@
+		return err
+	}
+
+	// pull the image from ecr. it will be public so no creds needed?
+	// out, err := cli.ImagePull(ctx, imageName, types.ImagePullOptions{})
+	// if err != nil {
+	// 	return err
+	// }
+	// defer out.Close()
+	// io.Copy(os.Stdout, out)
+	// log.CliLogger.Tracef("Pull confluent-local image success")
```
```suggestion
```

MuweiHe replied:
this will be added back when image becomes public, to pull the image first.

MuweiHe commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,161 @@
+		return err
+	}
+
+	// pull the image from ecr. it will be public so no creds needed?
```
this will be added when the image is public

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,168 @@
+func (c *localCommand) newStartCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "start",
+		Short: "Start local Kafka service.",
```
```suggestion
		Short: "Start a single-node instance of Apache Kafka.",
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,168 @@
+		Image:        imageName,
+		Hostname:     "broker",
+		Cmd:          strslice.StrSlice{"bash", "-c", "'/etc/confluent/docker/run'"},
+		ExposedPorts: nat.PortSet{nat.Port(ports.RestPort + "/tcp"): struct{}{}, nat.Port(ports.PlaintextPort + "/tcp"): struct{}{}},
```
```suggestion
		ExposedPorts: nat.PortSet{
			nat.Port(ports.RestPort + "/tcp"): struct{}{},
			nat.Port(ports.PlaintextPort + "/tcp"): struct{}{},
		},
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,168 @@
+	}
+	hostConfig := &container.HostConfig{
+		PortBindings: nat.PortMap{
+			nat.Port(ports.RestPort + "/tcp"): []nat.PortBinding{
```
Let's store this in a variable? It's used multiple times.

MuweiHe replied:
yes!

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,168 @@
+		PortBindings: nat.PortMap{
+			nat.Port(ports.RestPort + "/tcp"): []nat.PortBinding{
+				{
+					HostIP:   "0.0.0.0",
```
Can we say "localhost"?

MuweiHe replied:
yes! changed it!

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,168 @@
+	if err != nil {
+		return err
+	}
+	log.CliLogger.Tracef("Start confluent-local container success")
```
Not needed since we print to stdout below.
```suggestion
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,168 @@
+	}
+	log.CliLogger.Tracef("Start confluent-local container success")
+
+	output.Printf("Started Confluent local container %v.\nContinue your experience with Confluent local running `confluent local kafka produce` and `confluent local kafka consume`.\n", createResp.ID[:10])
```
Why `:10`?

MuweiHe replied:
it's a convention i found online, people use :10 of the super long container id.

brianstrauch replied:
Let's do something like `const containerIdShortLength = 10`?

brianstrauch replied:
If someone reads this code in the future, they won't know what 10 means.

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,168 @@
+
+func checkIsDockerRunning(dockerClient *client.Client) error {
+	_, err := dockerClient.Info(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
```
```suggestion
	return err
```

brianstrauch commented on internal/cmd/local/command_kafka_stop.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+func (c *localCommand) newStopCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "stop",
+		Short: "Stop the local Kafka service.",
+		Args:  cobra.NoArgs,
+		RunE:  c.stop,
+	}
+
+	return cmd
```
```suggestion
	return &cobra.Command{
		Use:   "stop",
		Short: "Stop the local Apache Kafka service.",
		Args:  cobra.NoArgs,
		RunE:  c.stop,
	}
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,168 @@
+
+	output.Printf("Started Confluent local container %v.\nContinue your experience with Confluent local running `confluent local kafka produce` and `confluent local kafka consume`.\n", createResp.ID[:10])
+	return nil
+}
```
This function is about 86 lines long. Can we make it shorter? The configuration seems like an easy thing to break out and unit test separately.

brianstrauch commented on internal/pkg/errors/strings.go:
```diff
@@ -60,13 +60,7 @@ const (
-	StoppingServiceMsg         = "Stopping %s\n"
-	ServiceStatusMsg           = "%s is [%s]\n"
-	DestroyDeletingMsg         = "Deleting: %s\n"
+	ConfluentLocalThankYouMsg = "Thank you for using Confluent local.\n"
```
This seems a bit cheesy, what do you think?

MuweiHe replied:
idk lol i wanted it to look like some friendly trial?

brianstrauch commented on internal/pkg/errors/error_message.go:
```diff
@@ -209,18 +209,13 @@ const (
+	ConfluentLocalStartedErrorMsg    = "Confluent local container has been created"
+	ConfluentLocalStartedSuggestions = "Continue your experience with Confluent local running `confluent local kafka produce` and `confluent local kafka consume`.\n" +
+		"To stop Confluent local container, run `confluent local kafka stop`."
+	SavePortsToConfigErrorMsg     = "failed to save local ports to config"
```
My philosophy on error messages is:
1. If it's used once, it should be put in-line.
2. If it's used multiple times in the same package, it should be declared in that package.
3. If it's used across multiple packages, it should be declared here.

I think the overall idea is that limiting the scope of variables/functions reduces the chance of conflicts and the need to create long variable names.

MuweiHe replied:
thanks! Modified some of them

brianstrauch commented on test/fixtures/output/context/create/0.golden:
```diff
@@ -1,5 +1,5 @@
 +------------+-----------------------+
 | Name       |                     0 |
-| Platform   | http://127.0.0.1:1024 |
+| Platform   | http://127.0.0.1:1026 |
```
Why is this being changed?

MuweiHe replied:
reverted! maybe i was already running a docker instance?

brianstrauch commented on test/local_test.go:
```diff
@@ -1,147 +1,15 @@
-	})
-	defer s.destroy()
-
+func (s *CLITestSuite) TestLocal() {
```
Make sure to test `local kafka topic`!

brianstrauch commented on test/fixtures/output/local/topic_delete.golden:
```diff
@@ -0,0 +1,4 @@
+The local commands are intended for a single-node development environment only, NOT for production usage. See more: https://docs.confluent.io/current/cli/index.html
+As of Confluent Platform 8.0, Java 8 is no longer supported.
```
```suggestion
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,171 @@
+		return err
+	}
+
+	output.Printf("Started Confluent local container %v.\nContinue your experience with Confluent local running `confluent local kafka produce` and `confluent local kafka consume`.\n", createResp.ID[:containerIdShortLength])
```
Instead of a global variable, I'd recommend a function that returns the shortened container ID. That way, you can declare the constant inside the function and limit it's scope.

MuweiHe replied:
`start` gets the id from resp and `stop` gets in from a container object

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,171 @@
+		return err
+	}
+
+	output.Printf("Started Confluent local container %v.\nContinue your experience with Confluent local running `confluent local kafka produce` and `confluent local kafka consume`.\n", createResp.ID[:containerIdShortLength])
```
```suggestion
	output.Printf("Started Confluent local container %v.\nContinue your experience with Confluent local running `confluent local kafka topic produce` and `confluent local kafka topic consume`.\n", createResp.ID[:containerIdShortLength])
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,171 @@
+		return nil
+	}
+
+	restPortInt, err := freeport.GetFreePort()
```
Ports are always integers, so this is probably a bit redundant. Also "rest" means Kafka REST, right?
```suggestion
	kafkaRestPort, err := freeport.GetFreePort()
```

brianstrauch commented on internal/cmd/local/command_kafka_stop.go:
```diff
@@ -0,0 +1,63 @@
+	if err := c.Config.Save(); err != nil {
+		return errors.Wrap(err, "failed to remove local ports from config")
+	}
+
```
```suggestion
```

brianstrauch commented on internal/cmd/local/command_kafka_topic.go:
```diff
@@ -0,0 +1,63 @@
+		Args:  cobra.NoArgs,
+	}
+
+	cmd.AddCommand(c.newListCommand())
+	cmd.AddCommand(c.newCreateCommand())
+	cmd.AddCommand(c.newDescribeCommand())
+	cmd.AddCommand(c.newDeleteCommand())
+	cmd.AddCommand(c.newUpdateCommand())
+	cmd.AddCommand(c.newProduceCommand())
+	cmd.AddCommand(c.newConsumeCommand())
```
Please sort alphabetically!

brianstrauch commented on internal/cmd/local/command_kafka_topic.go:
```diff
@@ -0,0 +1,63 @@
+		return nil, "", err
+	}
+
+	kafkaREST := pcmd.KafkaREST{
```
```suggestion
	kafkaRest := pcmd.KafkaREST{
```

brianstrauch commented on internal/cmd/local/command_kafka_topic_consume.go:
```diff
@@ -0,0 +1,163 @@
+	"github.com/confluentinc/cli/internal/pkg/output"
+)
+
+func (c *command) newConsumeCommand() *cobra.Command {
```
Since `command` exists at the top-level, we should format like this to avoid naming conflicts in the future.
```suggestion
func (c *command) newKafkaTopicConsumeCommand() *cobra.Command {
```

brianstrauch commented on internal/cmd/local/command_kafka_topic_consume.go:
```diff
@@ -0,0 +1,163 @@
+	cmd := &cobra.Command{
+		Use:   "consume <topic>",
+		Args:  cobra.ExactArgs(1),
+		RunE:  c.topicConsume,
```
```suggestion
		RunE:  c.kafkaTopicConsume,
```

brianstrauch commented on internal/cmd/local/command_kafka_topic_consume.go:
```diff
@@ -0,0 +1,163 @@
+		Long:  "Consume messages from a Kafka topic. Configuration and command guide: https://docs.confluent.io/confluent-cli/current/cp-produce-consume.html.\n\nTruncated message headers will be printed if they exist.",
+		Example: examples.BuildExampleString(
+			examples.Example{
+				Text: `Consume message from topic "my_topic" from the beginning and with keys printed.`,
```
Nit: Can we avoid mixing symbols in our examples, and just stick with lower-case-with-dashes?
```suggestion
				Text: `Consume message from topic "my-topic" from the beginning and with keys printed.`,
```

brianstrauch commented on internal/cmd/local/command_kafka_topic_consume.go:
```diff
@@ -0,0 +1,163 @@
+		Example: examples.BuildExampleString(
+			examples.Example{
+				Text: `Consume message from topic "my_topic" from the beginning and with keys printed.`,
+				Code: `confluent local kafka topic consume my_topic --from-beginning --print-key`,
```
```suggestion
				Code: "confluent local kafka topic consume my_topic --from-beginning --print-key",
```

brianstrauch commented on internal/cmd/local/command_kafka_topic_consume.go:
```diff
@@ -0,0 +1,163 @@
+	cmd.Flags().Bool("timestamp", false, "Print message timestamp in milliseconds.")
+	cmd.Flags().String("delimiter", "\t", "The delimiter separating each key and value.")
+	cmd.Flags().StringSlice("config", nil, `A comma-separated list of configuration overrides ("key=value") for the consumer client.`)
+	cmd.Flags().String("config-file", "", "The path to the configuration file (in json or avro format) for the consumer client.")
```
```suggestion
	cmd.Flags().String("config-file", "", "The path to the configuration file (in JSON or Avro format) for the consumer client.")

```

brianstrauch replied:
Can you also add `MarkFilename()` for this flag so autocompletion works a little better?

brianstrauch commented on internal/cmd/local/command_kafka_topic_create.go:
```diff
@@ -0,0 +1,43 @@
+func (c *command) topicCreate(cmd *cobra.Command, args []string) error {
+	restClient, clusterId, err := initKafkaRest(c.CLICommand, cmd)
+	if err != nil {
+		fmt.Println("err", err)
```
```suggestion
```

brianstrauch commented on internal/cmd/local/command_kafka_topic_list.go:
```diff
@@ -0,0 +1,31 @@
+		RunE:  c.topicList,
+	}
+
+	pcmd.AddOutputFlag(cmd)
```
```suggestion
	pcmd.AddOutputFlag(cmd)
	
```

brianstrauch commented on internal/cmd/local/command_kafka_topic_produce.go:
```diff
@@ -0,0 +1,170 @@
+	scanner := bufio.NewScanner(os.Stdin)
+	// On-prem Kafka messageMaxBytes: using the same value of cloud. TODO: allow larger sizes if customers request
+	// https://github.com/confluentinc/cc-spec-kafka/blob/9f0af828d20e9339aeab6991f32d8355eb3f0776/plugins/kafka/kafka.go#L43.
+	const maxScanTokenSize = 1024*1024*2 + 12
+	scanner.Buffer(nil, maxScanTokenSize)
+	input := make(chan string, 1)
+	// Avoid blocking in for loop so ^C or ^D can exit immediately.
+	var scanErr error
+	scan := func() {
+		hasNext := scanner.Scan()
+		if !hasNext {
+			// Actual error.
+			if scanner.Err() != nil {
+				scanErr = scanner.Err()
+			}
+			// Otherwise just EOF.
+			close(input)
+		} else {
+			input <- scanner.Text()
+		}
+	}
```
Any way we can avoid duplicating this code? If we change this functionality in the future, it would be great not to have to change it in two places.

brianstrauch commented on internal/cmd/local/command_kafka_topic_produce.go:
```diff
@@ -0,0 +1,170 @@
+			continue
+		}
+
+		msg, err := kafka.GetProduceMessage(cmd, []byte{0, 0, 0, 0}, topicName, data, serializationProvider)
```
```suggestion
		msg, err := kafka.GetProduceMessage(cmd, make([]byte, 4), topicName, data, serializationProvider)
```

brianstrauch commented on test/local_test.go:
```diff
@@ -1,147 +1,31 @@
 
 	for _, tt := range tests {
 		tt.workflow = true
+		tt.login = "platform"
```
This shouldn't require CP login!

brianstrauch commented on test/local_test.go:
```diff
@@ -1,147 +1,29 @@
-	defer s.destroy()
+import "time"
 
+func (s *CLITestSuite) TestLocal() {
```
```suggestion
func (s *CLITestSuite) TestLocalKafka() {
```

brianstrauch commented on internal/pkg/errors/error_message.go:
```diff
@@ -209,7 +209,7 @@ const (
 
 	// local commands
-	ConfluentLocalStartedErrorMsg    = "Confluent local container has been created"
+	ConfluentLocalStartedErrorMsg    = "Confluent local container has already been created"
```
Nit: I assume the product is called "Confluent Local" (both words uppercase)
```suggestion
	ConfluentLocalStartedErrorMsg    = "Confluent Local container has already been created"
```

brianstrauch commented:
Looks good so far!

* Running `confluent local kafka stop` prints "Thank you for using Confluent Local." even when `confluent local kafka start` hasn't been run.
* Running `confluent local kafka topic create test` when Confluent Local hasn't been started suggests that I stop Confluent Local.
* The Docker output should be formatted like how Docker formats it!
* Suggestion: Add `confluent local start` which can run all of the components? It looks better if the user has to type a short command to get started.
* Suggestion: Only print "The local commands are intended for a single-node..." when a `start` command is run. Printing it for every command is a bit annoying.

MuweiHe commented:
Need some help with release notes

sgagniere commented:
Suggestion: in,

Remove inefficient flag `--output` in `confluent kafka topic consume` for on-prem CLI

replace "inefficient" with "unused".

MuweiHe commented:
> Looks good so far!
> 
> * Running `confluent local kafka stop` prints "Thank you for using Confluent Local." even when `confluent local kafka start` hasn't been run.
Addressed this!

> * Running `confluent local kafka topic create test` when Confluent Local hasn't been started suggests that I stop Confluent Local.
i wrote this whenever reading ports from config fails, like config.json was corrupted could lead to this msg as well. Stopping it first would be a safe measure to ensure the container was removed first

> * The Docker output should be formatted like how Docker formats it!
I'm getting the output from Docker's official cli client. couldn't change the format it gives me.

> * Suggestion: Add `confluent local start` which can run all of the components? It looks better if the user has to type a short command to get started.
For the purpose of releasing this by eod i will address this issue in a following pr, when the next component is ready!

> * Suggestion: Only print "The local commands are intended for a single-node..." when a `start` command is run. Printing it for every command is a bit annoying.
Addressed this. Thanks!

brianstrauch commented:
> I'm getting the output from Docker's official cli client. couldn't change the format it gives me.

What I mean by this is we should try to print the status field from the JSON, not the whole JSON response. Instead of this:
```
{"status":"Pulling from confluentinc/confluent-local","id":"latest"}
{"status":"Digest: sha256:1aeeebdef8e20548c36ef23df8e10e685b9e4ac2b268b36b045d8d2e3e71c0d9"}
{"status":"Status: Image is up to date for confluentinc/confluent-local:latest"}
```
Something like this:
```
latest: Pulling from confluentinc/confluent-local
Digest: sha256:1aeeebdef8e20548c36ef23df8e10e685b9e4ac2b268b36b045d8d2e3e71c0d9
Status: Image is up to date for confluentinc/confluent-local:latest
```

tl;dr, we should try to match the output of `docker pull confluentinc/confluent-local`.

brianstrauch commented:
> i wrote this whenever reading ports from config fails, like config.json was corrupted could lead to this msg as well. Stopping it first would be a safe measure to ensure the container was removed first

That makes sense, but is there any way to tell those two cases apart? (The container is stopped vs. we can't read the ports from the configuration file)

brianstrauch commented on cmd/lint/main.go:
```diff
@@ -19,10 +19,8 @@ var commandRules = []linter.CommandRule{
-		linter.ExcludeCommandContains("local services"),
 		linter.ExcludeCommand("kafka client-config create nodejs")),
 
-	linter.Filter(linter.RequireCapitalizeProperNouns("Short", properNouns), linter.ExcludeCommand("local current")),
```
Can't suggest an edit here, but you're deleting the entire rule. It should be rewritten to the following instead of deleted:
```
linter.RequireCapitalizeProperNouns("Short", properNouns),
```

brianstrauch commented on internal/cmd/kafka/command_topic_consume.go:
```diff
@@ -51,7 +51,7 @@ func newConsumeCommand(prerunner pcmd.PreRunner, clientId string) *cobra.Command
 	cmd.Flags().Bool("timestamp", false, "Print message timestamp in milliseconds.")
 	cmd.Flags().StringSlice("config", nil, `A comma-separated list of configuration overrides ("key=value") for the consumer client.`)
-	cmd.Flags().String("config-file", "", "The path to the configuration file (in json or avro format) for the consumer client.")
+	cmd.Flags().String("config-file", "", "The path to the configuration file (in JSON or avro format) for the consumer client.")
```
```suggestion
	cmd.Flags().String("config-file", "", "The path to the configuration file for the consumer client, in JSON or Avro format.")
```

brianstrauch commented on internal/cmd/kafka/command_topic_consume_onprem.go:
```diff
@@ -50,9 +50,8 @@ func (c *authenticatedTopicCommand) newConsumeCommandOnPrem() *cobra.Command {
 	cmd.Flags().String("delimiter", "\t", "The delimiter separating each key and value.")
 	cmd.Flags().StringSlice("config", nil, `A comma-separated list of configuration overrides ("key=value") for the consumer client.`)
-	cmd.Flags().String("config-file", "", "The path to the configuration file (in json or avro format) for the consumer client.")
+	cmd.Flags().String("config-file", "", "The path to the configuration file (in JSON or avro format) for the consumer client.")
```
This flag is defined more than once. Can we write an `addConfigFile()` helper function?

MuweiHe replied:
there're 4 kinds of different definitions of `config-file` flags in multiple commands. I'll do a clean-up pr to address all of them

brianstrauch commented on internal/cmd/kafka/command_topic_create_onprem.go:
```diff
@@ -62,6 +63,10 @@ func (c *authenticatedTopicCommand) createOnPrem(cmd *cobra.Command, args []stri
 		return err
 	}
 
+	return CreateTopicWithRestClient(cmd, restClient, restContext, topicName, clusterId)
```
```suggestion
	return CreateTopic(cmd, restClient, restContext, topicName, clusterId)
```

brianstrauch commented on internal/cmd/kafka/confluent_kafka.go:
```diff
@@ -246,7 +246,7 @@ func consumeMessage(e *ckafka.Message, h *GroupHandler) error {
 }
 
-func runConsumer(consumer *ckafka.Consumer, groupHandler *GroupHandler) error {
+func RunConsumer(consumer *ckafka.Consumer, groupHandler *GroupHandler) error {
```
All public functions shared across multiple packages should probably get moved to `internal/pkg/kafka` to avoid future import cycle issues.

MuweiHe replied:
tried it and brought many import cycles from cmd, sr, and kafka. Will do that all in the clean-up pr as well.

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+const (
+	confluentLocalImageName     = "confluentinc/confluent-local:latest"
```
We're inside the Confluent Local package, so we don't need to prefix everything with "Confluent Local", but we should probably be clear that this is for Docker.
```suggestion
	dockerImageName     = "confluentinc/confluent-local:latest"
```

MuweiHe replied:
good point!

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,45 @@
+	confluentLocalContainerName = "confluent-local"
+	localhostPrefix             = "http://localhost:%s"
+	localhost                   = "localhost"
+	kafkaRESTNotReadySuggestion = "Kafka REST connection may not be ready yet. Rerunning the command is suppposed to solve the issue."
```
```suggestion
	kafkaRestNotReadySuggestion = "Kafka REST connection is not ready. Re-running the command may solve the issue."
```

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,45 @@
+
+	cmd.AddCommand(c.newKafkaStartCommand())
+	cmd.AddCommand(c.newKafkaStopCommand())
+	cmd.AddCommand(c.newTopicCommand())
```
```suggestion
	cmd.AddCommand(c.newKafkaTopicCommand())
```

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+func getShortenedContainerId(id string) string {
+	containerIdShortLength := 10
+	return id[:containerIdShortLength]
```
```suggestion
	return id[:10]
```

MuweiHe replied:
could do this but in earlier comments it was suggested a single `10` is not very clear what it means

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,170 @@
+		return err
+	}
+	defer out.Close()
+	_, err = io.Copy(os.Stdout, out)
+	if err != nil {
+		return err
+	}
```
I think this should be equivalent?
```suggestion
	output.Print(string(out))
```

MuweiHe replied:
will try to marshal json output

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,170 @@
+	}
+	log.CliLogger.Tracef("Pull confluent-local image success")
+
+	err = c.prepareAndSaveLocalPorts()
+	if err != nil {
```
```suggestion
	if err := c.prepareAndSaveLocalPorts(); err != nil {
```

MuweiHe commented:
> > i wrote this whenever reading ports from config fails, like config.json was corrupted could lead to this msg as well. Stopping it first would be a safe measure to ensure the container was removed first
> 
> That makes sense, but is there any way to tell those two cases apart? (The container is stopped vs. we can't read the ports from the configuration file)

That would need extra calls to docker client in every command?

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,170 @@
+		return err
+	}
+
+	output.Printf("Started Confluent Local container %v.\nContinue your experience with Confluent Local running `confluent local kafka topic create test` and `confluent local kafka topic produce test`.\n", getShortenedContainerId(createResp.ID))
```
```suggestion
	output.Printf("Started Confluent Local container %v.\nTo continue your Confluent Local experience, run `confluent local kafka topic create test` and `confluent local kafka topic produce test`.\n", getShortenedContainerId(createResp.ID))
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -0,0 +1,170 @@
+		},
+		Env: getContainerEnvironmentWithPorts(ports),
+	}
+	hostConfig := &container.HostConfig{
+		PortBindings: nat.PortMap{
+			natKafkaRestPort: []nat.PortBinding{
+				{
+					HostIP:   localhost,
+					HostPort: ports.KafkaRestPort,
+				},
+			},
+			natPlaintextPort: []nat.PortBinding{
+				{
+					HostIP:   localhost,
+					HostPort: ports.PlaintextPort,
+				},
+			},
+		},
+	}
```
There's only one field, so we can remove some indentation here.
```suggestion
	hostConfig := &container.HostConfig{PortBindings: nat.PortMap{
		natKafkaRestPort: []nat.PortBinding{
			{
				HostIP:   localhost,
				HostPort: ports.KafkaRestPort,
			},
		},
		natPlaintextPort: []nat.PortBinding{
			{
				HostIP:   localhost,
				HostPort: ports.PlaintextPort,
			},
		},
	}}
```

brianstrauch commented on internal/cmd/local/command_kafka_stop.go:
```diff
@@ -0,0 +1,64 @@
+			}
+			log.CliLogger.Tracef("Confluent Local container removed")
+
+			output.Printf(errors.ConfluentLocalThankYouMsg)
```
What about the following?
```
Stopped Confluent Local.
```

brianstrauch replied:
I believe we say "Started Confluent Local." after `confluent local kafka start`?

MuweiHe commented on .semaphore/semaphore.yml:
```diff
@@ -18,6 +18,9 @@ blocks:
     dependencies: []
     task:
       jobs:
+        - name: Docker check
+          commands:
+            - docker ps
```
Not sure i understand this: this won't fail, it'll just start docker socket.

brianstrauch replied:
Oh? Running `docker ps` starts docker? I thought that's how you printed the list of containers

MuweiHe replied:
actually the daemon is already running in semaphore so i deleted this! Thought need to trigger a docker command first.:(

brianstrauch commented on test/fixtures/output/local/start.golden:
```diff
@@ -1,7 +1,7 @@
-{"status":"Status: Image is up to date for confluentinc/confluent-local:latest"}
+Pulling from confluentinc/confluent-local","id":"latest"
+Digest: sha256:[a-zA-Z0-9]*.
+Status: Image is up to date for confluentinc/confluent-local:Latest
```
```suggestion
Status: Image is up to date for confluentinc/confluent-local:latest
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -65,7 +65,9 @@ func (c *command) kafkaStart(cmd *cobra.Command, args []string) error {
 			continue
 		}
-		fmt.Printf("Status\t%v\n", output.Status)
+		if len(output.Status) > 18 { // filter out lines such as "Extracing" and "Pulling"
```
Can we blacklist these lines instead of using an arbitrary length like 18?
